File: agentic_output_FullAdder1Bit_chisel_qwen2.5-coder_7b_20260810_232120/src/main/scala/mxfp4/MXFP4.scala
package mxfp4

import chisel3._
import chisel3.experimental.BundleLiterals._

// Formato MXFP4 E2M1 (OCP MX Specification v1.0): 4 bit totali.
//   bit[3]   = segno      (0 = positivo, 1 = negativo)
//   bit[2:1] = esponente a 2 bit (bias = 1)
//   bit[0]   = mantissa a 1 bit
// Valore: (-1)^sign * 2^(exp-1) * (1 + mant*0.5)
class MXFP4 extends Bundle {
  val sign = Bool()
  val exp  = UInt(2.W)
  val mant = UInt(1.W)
}

object MXFP4 {
  def apply(): MXFP4 = new MXFP4

  def apply(bits: Int): MXFP4 = {
    val s = (bits >> 3) & 0x1
    val e = (bits >> 1) & 0x3
    val m = bits & 0x1
    (new MXFP4).Lit(_.sign -> (s == 1).B, _.exp -> e.U(2.W), _.mant -> m.U(1.W))
  }

  def decode(bits: Int): Double = {
    val sign = (bits >> 3) & 0x1
    val exp  = (bits >> 1) & 0x3
    val mant = bits & 0x1
    val s = if (sign == 1) -1.0 else 1.0
    if (exp == 0) {
      s * mant * 0.5
    } else {
      s * (1.0 + mant * 0.5) * math.pow(2.0, exp - 1)
    }
  }

  def encode(value: Double): Int = {
    var bestBits = 0
    var bestDiff = Double.MaxValue
    for (bits <- 0 until 16) {
      val diff = math.abs(decode(bits) - value)
      if (diff < bestDiff) {
        bestDiff = diff
        bestBits = bits
      }
    }
    bestBits
  }
}

File: agentic_output_FullAdder_amaranth_qwen2.5-coder_latest_20260810_090745/mxfp4.py
"""Formato MXFP4 E2M1 (OCP MX Specification v1.0): 4 bit totali.

  bit[3]   = segno      (0 = positivo, 1 = negativo)
  bit[2:1] = esponente a 2 bit (bias = 1)
  bit[0]   = mantissa a 1 bit

Valore: (-1)^segno * 2^(exp-1) * (1 + mant*0.5), tranne il caso subnormale
(exp == 0, include lo zero) dove NON c'e' bit implicito:
valore = (-1)^segno * mant*0.5.
"""
from amaranth.lib import data


class MXFP4Layout(data.Struct):
    # Ordine di dichiarazione = ordine dei bit da LSB a MSB in
    # amaranth.lib.data.Struct: mant (bit 0), exp (bit 2:1), sign (bit 3).
    mant: 1
    exp: 2
    sign: 1


def decode(bits: int) -> float:
    """Converte i 4 bit codificati (0..15) nel valore Double rappresentato."""
    sign = (bits >> 3) & 0x1
    exp  = (bits >> 1) & 0x3
    mant = bits & 0x1
    s = -1.0 if sign == 1 else 1.0
    if exp == 0:
        return s * mant * 0.5
    return s * (1.0 + mant * 0.5) * (2.0 ** (exp - 1))


def encode(value: float) -> int:
    """Converte un valore Double nella codifica MXFP4 (0..15) piu' vicina
    (round-to-nearest)."""
    best_bits, best_diff = 0, float("inf")
    for bits in range(16):
        diff = abs(decode(bits) - value)
        if diff < best_diff:
            best_diff, best_bits = diff, bits
    return best_bits
